helios-authentication/src/main/java/com/spotify/helios/authentication/AuthProviderLoader.java

```diff
@@ -0,0 +1,198 @@
+   *
+   * @param plugin      The plugin jar file to load.
+   * @param environment The class loader to use for providing plugin interface dependencies.
+   * @param parent      The parent class loader to assign to the class loader of the jar.
+   * @return A {@link AuthProviderFactory}, null if none could be found.
+   * @throws AuthProviderLoadingException if loading failed.
+   */
+  public static AuthProviderFactory load(final Path plugin,
+                                               final ClassLoader environment,
+                                               final ClassLoader parent)
+      throws AuthProviderLoadingException {
+    return load("plugin jar file: " + plugin, pluginClassLoader(plugin, environment, parent));
+  }
+
+  /**
+   * Load a {@link AuthProviderFactory} using a class loader.
+   *
+   * @param source      The source of the class loader.
+   * @param classLoader The class loader to load from.
+   * @return A {@link AuthProviderFactory}, null if none could be found.
+   * @throws AuthProviderLoadingException if loading failed.
+   */
+  public static AuthProviderFactory load(final String source, final ClassLoader classLoader)
+      throws AuthProviderLoadingException {
+    final ServiceLoader<AuthProviderFactory> loader;
+    try {
+      loader = ServiceLoader.load(AuthProviderFactory.class, classLoader);
+    } catch (ServiceConfigurationError e) {
+      throw new AuthProviderLoadingException(
+          "Failed to load service registrar from " + source, e);
+    }
+    final Iterator<AuthProviderFactory> iterator = loader.iterator();
+    if (iterator.hasNext()) {
+      return iterator.next();
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Attempt to get the extension class loader, which can only load jdk classes and classes from
+   * jvm
+   * extensions. Adapted from {@link ServiceLoader#loadInstalled(Class)}
+   */
+  private static ClassLoader extensionClassLoader(final ClassLoader environment) {
+    ClassLoader cl = environment;
+    ClassLoader prev = null;
+    while (cl != null) {
+      prev = cl;
+      cl = cl.getParent();
+    }
+    return prev;
+  }
+
+  /**
+   * Create a class loader for a plugin jar.
+   */
+  private static ClassLoader pluginClassLoader(final Path plugin,
+                                               final ClassLoader environment,
+                                               final ClassLoader parent) {
+    final URL url;
+    try {
+      url = plugin.toFile().toURI().toURL();
+    } catch (MalformedURLException e) {
+      throw new RuntimeException("Failed to load plugin jar " + plugin, e);
+    }
+    final ClassLoader providedClassLoader = new FilteringClassLoader(PROVIDED, environment, parent);
+    return new URLClassLoader(new URL[]{url}, providedClassLoader);
+  }
+
+  /**
+   * A class loader that exposes a specific list of packages from another class loader.
+   */
+  private static class FilteringClassLoader extends ClassLoader {
```

Comment:
this seems copied verbatim from `com.spotify.helios.serviceregistration.ServiceRegistrarLoader`, would be nice to not have to duplicate those